File: python/2020/june/june-22/main.py
import math

def gcd(a, b):
  c = a % b
  while(c > 0):
    a = b
    b = c
    c = a % b
  return b


def greedy_egyptian_fraction(num, den, denominators = [], index = 0):
  if(num == 1):
    denominators.append(den)
  else:
    unit_den = math.ceil(den / num)
    denominators.append(unit_den)
    gcd_of_numbers = gcd((num * unit_den) - den, den * unit_den)
    greedy_egyptian_fraction(((num * unit_den) - den) // gcd_of_numbers, (den * unit_den) // gcd_of_numbers, denominators, index + 1)
  
  return denominators


if __name__ == '__main__':
  result = greedy_egyptian_fraction(4, 13)
  print("Egyptian fraction of 4 / 13")

  for num in result:
    print(f"1 / {num}")

File: rust/2020/june/june-22/src/main.rs
// * Daily Coding Problem June 22 2020

// * [Easy] -- Palantir

// * The ancient Egyptians used to express fractions as a sum of several terms where each numerator is one.
// ! Ex: 4 / 13 can be represented as 1 / 4 + 1 / 18 + 1 / 468.

// * Create an algorithm to turn an ordinary fraction a / b, where a < b, into an Egyptian fraction.

// * This will return an array of the denominators in Egyptian form since all numerators will be 1

use math::round;

fn egyptian_fraction(num: u32, den: u32, mut denoms: Vec<u32>, idx: usize) -> Vec<u32> {
    if num == 1 {
        denoms.push(den);
    } else {
        let next_denom = round::ceil((den / num).into(), 0) as u32;
        denoms.push(next_denom);
        let gcd = greatest_common_divisor(num * next_denom - den, den * next_denom);
        return egyptian_fraction(round::floor((num * next_denom - den) / gcd, 0), round);
    }
    return denoms;
}

fn main() {
    println!("Egyptian fraction of 4 / 13");
}
